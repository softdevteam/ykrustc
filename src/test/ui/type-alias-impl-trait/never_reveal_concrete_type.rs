// revisions: min_tait full_tait
#![feature(min_type_alias_impl_trait)]
#![cfg_attr(full_tait, feature(type_alias_impl_trait))]
//[full_tait]~^ WARN incomplete

fn main() {}

// don't reveal the concrete type
type NoReveal = impl std::fmt::Debug;

fn define_no_reveal() -> NoReveal {
    ""
}

fn no_reveal(x: NoReveal) {
    let _: &'static str = x; //~ mismatched types
    let _ = x as &'static str; //~ non-primitive cast
}
