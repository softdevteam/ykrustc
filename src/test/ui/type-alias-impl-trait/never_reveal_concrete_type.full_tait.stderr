warning: the feature `type_alias_impl_trait` is incomplete and may not be safe to use and/or cause compiler crashes
  --> $DIR/never_reveal_concrete_type.rs:3:32
   |
LL | #![cfg_attr(full_tait, feature(type_alias_impl_trait))]
   |                                ^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(incomplete_features)]` on by default
   = note: see issue #63063 <https://github.com/rust-lang/rust/issues/63063> for more information

error[E0308]: mismatched types
  --> $DIR/never_reveal_concrete_type.rs:16:27
   |
LL | type NoReveal = impl std::fmt::Debug;
   |                 -------------------- the found opaque type
...
LL |     let _: &'static str = x;
   |            ------------   ^ expected `&str`, found opaque type
   |            |
   |            expected due to this
   |
   = note: expected reference `&'static str`
            found opaque type `impl Debug`

error[E0605]: non-primitive cast: `impl Debug` as `&'static str`
  --> $DIR/never_reveal_concrete_type.rs:17:13
   |
LL |     let _ = x as &'static str;
   |             ^^^^^^^^^^^^^^^^^ an `as` expression can only be used to convert between primitive types or to coerce to a specific trait object

error: aborting due to 2 previous errors; 1 warning emitted

Some errors have detailed explanations: E0308, E0605.
For more information about an error, try `rustc --explain E0308`.
